src/plugins/core/cell.ts

```diff
@@ -618,25 +617,46 @@ export class CellPlugin extends CorePlugin<CoreState> implements CoreState {
   }
 }
 
-class FormulaCellWithDependencies {
+class FormulaCellWithDependencies implements FormulaCell {
   readonly isFormula = true;
+  readonly compiledFormula: RangeCompiledFormula;
   constructor(
     readonly id: UID,
-    readonly compiledFormula: CompiledFormula,
+    compiledFormula: CompiledFormula,
     readonly format: Format | undefined,
     readonly style: Style | undefined,
-    readonly dependencies: Range[],
+    dependencies: Range[],
     private readonly sheetId: UID,
-    private readonly getFormulaCellContent: (
-      sheetId: UID,
-      cell: Pick<FormulaCell, "dependencies" | "compiledFormula">
-    ) => string
-  ) {}
+    private readonly getRangeString: (range: Range, sheetId: UID) => string
+  ) {
+    let rangeIndex = 0;
+    const tokens = compiledFormula.tokens.map((token) => {
+      if (token.type === "REFERENCE") {
+        const index = rangeIndex++;
+        return new RangeReferenceToken(() =>
+          this.getRangeString(dependencies[index], this.sheetId)
+        );
+      }
+      return token;
+    });
+    this.compiledFormula = {
+      ...compiledFormula,
+      dependencies,
+      tokens,
+    };
+  }
 
   get content() {
-    return this.getFormulaCellContent(this.sheetId, {
-      dependencies: this.dependencies,
-      compiledFormula: this.compiledFormula,
-    });
+    return this.compiledFormula.tokens.map((token) => token.value).join("");
```

Comment:
you can use `concat`, it's supposed to be faster

Reply:
Yeah I saw it afterward, I added it in one of the push force :)

Reply:
push force was push forced ? :smile: